### PR TITLE
Scalar operation compilation

### PR DESCRIPTION
**Pull request by rachitnigam**

### CODE REVIEW

**Comment by rachitnigam** [comment]
For input program:
```
(VecMAC
     (Vec4
      0
      (* (Get I 0) (Get F 2))
      (+
       (* (Get I 1) (Get F 2))
       (+
        (* (Get F 1) (Get I 4))
        (* (Get F 0) (Get I 5))))
      (+
       (* (Get I 2) (Get F 2))
       (+
        (* (Get F 1) (Get I 5))
        (* (Get F 0) (Get I 6)))))
     (LitVec4 0 0 0 0)
     (LitVec4 0 0 0 0))
```

this generates:

```
(list
 (let-bind 'let_0 'z_0 'float)
 (array-get 'get_0 'I 0)
 (array-get 'get_1 'F 2)
 (scalar-binop 'op_0 '* 'get_0 'get_1)
 (array-get 'get_2 'I 1)
 (array-get 'get_3 'F 2)
 (scalar-binop 'op_1 '* 'get_2 'get_3)
 (array-get 'get_4 'F 1)
 (array-get 'get_5 'I 4)
 (scalar-binop 'op_2 '* 'get_4 'get_5)
 (array-get 'get_6 'F 0)
 (array-get 'get_7 'I 5)
 (scalar-binop 'op_3 '* 'get_6 'get_7)
 (scalar-binop 'op_4 '+ 'op_2 'op_3)
 (scalar-binop 'op_5 '+ 'op_1 'op_4)
 (array-get 'get_8 'I 2)
 (array-get 'get_9 'F 2)
 (scalar-binop 'op_6 '* 'get_8 'get_9)
 (array-get 'get_10 'F 1)
 (array-get 'get_11 'I 5)
 (scalar-binop 'op_7 '* 'get_10 'get_11)
 (array-get 'get_12 'F 0)
 (array-get 'get_13 'I 6)
 (scalar-binop 'op_8 '* 'get_12 'get_13)
 (scalar-binop 'op_9 '+ 'op_7 'op_8)
 (scalar-binop 'op_10 '+ 'op_6 'op_9)
 (vec-lit 'lit_0 '(let_0 op_0 op_5 op_10) 'float)
 (vec-const 'shufs_0 '#(0 0 0 0) 'int)
 (vec-shuffle 'shuf-out_0 'shufs_0 '(Z))
 (vec-const 'shufs_1 '#(0 0 0 0) 'int)
 (vec-shuffle 'shuf-out_1 'shufs_1 '(Z))
 (vec-app 'mac-out_0 'vec-mac '(lit_0 shuf-out_0 shuf-out_1)))
```

**Comment by rachitnigam** [comment]
Second step done. Generates (for the same program as above):
```
int __attribute__((section(".dram0.data"))) Z[4] = {0, 0, 0, 0};
int __attribute__((section(".dram0.data"))) shufs_0[4] = {0, 0, 0, 0};
int __attribute__((section(".dram0.data"))) shufs_1[4] = {0, 0, 0, 0};
void kernel() {
float z_0 = 0;
  float get_0 = I[0];
  float get_1 = F[2];
  float op_0 = get_0 * get_1;
  float get_2 = I[1];
  float get_3 = F[2];
  float op_1 = get_2 * get_3;
  float get_4 = F[1];
  float get_5 = I[4];
  float op_2 = get_4 * get_5;
  float get_6 = F[0];
  float get_7 = I[5];
  float op_3 = get_6 * get_7;
  float op_4 = op_2 + op_3;
  float op_5 = op_1 + op_4;
  float get_8 = I[2];
  float get_9 = F[2];
  float op_6 = get_8 * get_9;
  float get_10 = F[1];
  float get_11 = I[5];
  float op_7 = get_10 * get_11;
  float get_12 = F[0];
  float get_13 = I[6];
  float op_8 = get_12 * get_13;
  float op_9 = op_7 + op_8;
  float op_10 = op_6 + op_9;
  float lit_0[4] = [z_0, op_0, op_5, op_10];
  xb_vecMxf32 shuf_out_0;
  shuf_out_0 = PDX_MOV_MXF32_FROM_MX32(PDX_SHFL_MX32(*((xb_vecMx32 *) Z), *((xb_vecMx32 *) shufs_0)));
  xb_vecMxf32 shuf_out_1;
  shuf_out_1 = PDX_MOV_MXF32_FROM_MX32(PDX_SHFL_MX32(*((xb_vecMx32 *) Z), *((xb_vecMx32 *) shufs_1)));
  xb_vecMxf32 mac_out_0 = lit_0;
  PDX_MULA_MXF32(mac_out_0, shuf_out_0, shuf_out_1);
}
```

**Comment by avanhatt** [comment]
Hm looks like something isn't quite right with the function codegen because this kernel should take arguments

File: src/egg-to-dios-dsl.rkt [hunk 1]
```diff
@@ -54,18 +33,45 @@
 ; Functional expression -> (name, list of DSL instructions)
 (define (egg-to-dios e)
   (match e
-    [(egg-get name idx) e]
+    [0
+     (define zero (new-name 'z))
```
Comment (avanhatt) [inline src/egg-to-dios-dsl.rkt:37]: I think this needs to be ``Z` to be consistent with the prelude, we can add it to some shared location

**Pull request merged**